### DISCUSSION TITLE
Builtin: provide builtin function similar to dir() in python

### DISCUSSION THREAD

**Discussion by lakca** [vlang/v#11859] [Ideas]
Provide builtin function similar to `dir()` in python to return all fields/properties/methods of struct/interface/value or exposes of module.

It will be benefit since V is still in beta, also friendly to users to quickly grasp the general of an object.

**Comment by larpon** [comment]
You can use `println` or `dump` for the fields
```v
struct Abc {
    def int
}

fn main() {
    a := Abc{}
    println(a)
    dump(a)
}
```

**Comment by lakca** [comment]
Thanks, This is not what expected.

For instance, I want to know what methods or properties the string type or a string value has, also what you are wrote does not work for module and builtins, and cannot return methods.

**Comment by dumblob** [comment]
Runtime introspection is AFAIK not implemented (and AFAIK it's not even on the *V 1.0* roadmap). Instead *V* offers a [language server](https://github.com/vlang/vls ) providing the information you're asking for (and if anything is missing in there - e.g. the "module functions/structs overview" if I'm not mistaken - please file a bug for it as it's a useful functionality).

Otherwise there is always an autogenerated documentation for modules.

-----

`dir()` in Python is a poor man's quasi-solution compared to what language server and documentation offer :wink:.

**Comment by lakca** [comment]
the VLS crashes too often to use (ast failed). Sometimes, just a lookup in repl is enough, since most is similar and idiomatic among programing languages.

**Comment by dumblob** [comment]
> the VLS crashes too often to use (ast failed).

Then it's a bug - could you please report it? It's important also for others (VLS becomes more and more important as *V* matures).

>Sometimes, just a lookup in repl is enough, since most is similar and idiomatic among programing languages.

Hm, I saw someone writing a dead-simple function fetching the documentation from internet and reformatting it (visually compressing it) - note this can't be done offline because you don't want to first fetch all existing modules to search through them. Such fn can then be triggered manually as well as automatically e.g. in case of an error in REPL. PRs welcome :wink:.

**Comment by lakca** [comment]
Hm, okay,  I still reserve my opinion. Maybe I think sometimes providing an alternative will be great.

In addition, the [VLS crash bug is reported](https://github.com/vlang/vls/issues/116)

**Comment by medvednikov** [comment]
I like the idea.

You can also do `v doc os` for an overview of all types and functions/methods.
